tests/testthat/test-TealData.R

```diff
@@ -1,238 +1,342 @@
+  df2 <- dataset("df2", df2, keys = "df2_id")
 
-  expect_silent(
-    constructor_wrapper(
-      join_keys = join_keys(join_key("df_1", "df_2", c("x" = "u")))
-    )$check_metadata()
+  join_keys1 <- join_keys(join_key("df1", "df1", "id"), join_key("df2", "df2", "df2_id"))
+  data <- TealData$new(df1, df2, join_keys = join_keys1)
+  # primary keys are not taken from datasets when calling TealData$new(), these are only added if using wrappers
+  testthat::expect_equal(
+    data$get_join_keys(),
+    join_keys1
   )
 
-  expect_silent(
-    constructor_wrapper(
-      join_keys = join_keys(join_key("df_1", "df_2", c("x" = "u", "y" = "v")))
-    )$check_metadata()
+  data2 <- TealData$new(df1, df2)
+  testthat::expect_equal(
+    data2$get_join_keys(),
+    join_keys()
   )
+})
 
-  expect_silent(
-    constructor_wrapper(
-      join_keys = join_keys(join_key("df_2", "df_2", c("u" = "u")))
-    )$check_metadata()
+testthat::test_that("copy(deep = TRUE) deep copies self and the items", {
+  test_ds0 <- TealDataset$new("test_ds0", head(mtcars), code = "test_ds0 <- head(mtcars)")
+  test_ds1 <- TealDatasetConnector$new(
+    dataname = "test_ds1",
+    pull_callable = CallableFunction$new(data.frame),
+    vars = list(test_ds0 = test_ds0)
   )
+  data <- TealData$new(test_ds0, test_ds1)
+  data_cloned <- data$copy(deep = TRUE)
+  testthat::expect_false(identical(data, data_cloned))
+  testthat::expect_false(identical(data_cloned$get_items()$test_ds0, test_ds0))
 })
 
-test_that("check_metadata fails if inconsistent join_keys for given datasets", {
-  df_1 <- data.frame(x = 1:10, y = 1:10)
-  df_2 <- data.frame(u = 1:10, v = 1:10)
-
-  constructor_wrapper <- function(join_keys) {
-    data <- teal_data(
-      dataset("df_1", df_1),
-      dataset("df_2", df_2),
-      join_keys = join_keys
-    )
-  }
-
-  expect_error(
-    constructor_wrapper(
-      join_keys = join_keys(join_key("df_1", "df_2", c("x" = "w")))
-    )$check_metadata()
+testthat::test_that("copy(deep = TRUE) keeps valid references between items", {
+  test_ds0 <- TealDataset$new("test_ds0", head(mtcars), code = "test_ds0 <- head(mtcars)")
+  test_ds1 <- TealDatasetConnector$new(
+    dataname = "test_ds1",
+    pull_callable = CallableFunction$new(data.frame),
+    vars = list(test_ds0 = test_ds0)
   )
+  data <- TealData$new(test_ds0, test_ds1)
+  data_cloned <- data$copy(deep = TRUE)
+  new_test_ds0 <- data_cloned$get_items()$test_ds0
+  new_test_ds1 <- data_cloned$get_items()$test_ds1
+  testthat::expect_identical(
+    new_test_ds1$get_var_r6()$test_ds0,
+    new_test_ds0
+  )
+})
 
-  expect_error(
-    constructor_wrapper(
-      join_keys = join_keys(join_key("df_1", "df_2", c("x" = "y", "v" = "v")))
-    )$check_metadata()
+testthat::test_that("TealData$print prints out expected output on basic input", {
+  adsl_raw <- as.data.frame(as.list(setNames(nm = get_cdisc_keys("ADSL"), object = list(1:3, letters[1:3]))))
+  adsl <- cdisc_dataset(
```

Comment:
unless you are explicitly testing something cdisc related I'd use dataset in tests

Reply:
done.